spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/annotation/TaskListenerExecutor.java

```diff
@@ -0,0 +1,246 @@
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.cloud.task.listener.TaskExecutionListener;
+import org.springframework.cloud.task.repository.TaskExecution;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.core.MethodIntrospector;
+import org.springframework.core.annotation.AnnotationUtils;
+
+/**
+ * Identifies all beans that contain a TaskExecutionListener annotation and stores the
+ * associated method so that it can be called by the {@link TaskExecutionListener} at the
+ * appropriate time.
+ *
+ * @author Glenn Renfro
+ */
+public class TaskListenerExecutor implements TaskExecutionListener{
+
+	private ConfigurableApplicationContext context;
+
+	private final static Logger logger = LoggerFactory.getLogger(TaskListenerExecutor.class);
+
+
+	private final Set<Class<?>> nonAnnotatedClasses =
+			Collections.newSetFromMap(new ConcurrentHashMap<Class<?>, Boolean>());
+
+	private Map<Method, BeforeTask> beforeTaskMethods;
+	private Map<Method, Object> beforeTaskInstances;
+
+	private Map<Method, AfterTask> afterTaskMethods;
+	private Map<Method, Object> afterTaskInstances;
+
+	private Map<Method, FailedTask> failedTaskMethods;
+	private Map<Method, Object> failedTaskInstances;
+
+	public TaskListenerExecutor(ConfigurableApplicationContext context){
+		this.context = context;
+		this.beforeTaskMethods = new HashMap<>();
+		this.beforeTaskInstances = new HashMap<>();
+		this.afterTaskMethods = new HashMap<>();
+		this.afterTaskInstances = new HashMap<>();
+		this.failedTaskMethods = new HashMap<>();
+		this.failedTaskInstances = new HashMap<>();
+		initializeExecutor();
+	}
+
+	/**
+	 * Executes all the methods that have been annotated with  &#064;BeforeTask.
+	 * @param taskExecution associated with the event.
+	 */
+	@Override
+	public void onTaskStartup(TaskExecution taskExecution) {
+		executeTaskListener(taskExecution, beforeTaskMethods.keySet(), beforeTaskInstances);
+	}
+
+	/**
+	 * Executes all the methods that have been annotated with  &#064;AfterTask.
+	 * @param taskExecution associated with the event.
+	 */
+	@Override
+	public void onTaskEnd(TaskExecution taskExecution) {
+		executeTaskListener(taskExecution, afterTaskMethods.keySet(), afterTaskInstances);
+	}
+
+	/**
+	 * Executes all the methods that have been annotated with  &#064;FailedTask.
+	 * @param throwable that was not caught for the task execution.
+	 * @param taskExecution associated with the event.
+	 */
+	@Override
+	public void onTaskFailed(TaskExecution taskExecution, Throwable throwable) {
+		executeTaskListenerWithThrowable(taskExecution, throwable,
+				failedTaskMethods.keySet(),failedTaskInstances);
+	}
+
+	private void executeTaskListener(TaskExecution taskExecution, Set<Method> methods, Map<Method, Object> instances){
+		for (Method method : methods) {
+			try {
+				method.invoke(instances.get(method),taskExecution);
```

Comment:
Does this require that the method signature accept the `TaskExecution` parameter?  Is it possible to make the parameter optional on the annotation?